datumaro/components/media.py

```diff
@@ -217,49 +247,171 @@ def size(self) -> Optional[Tuple[int, int]]:
+        self._ext = self._ext if self._ext else osp.splitext(osp.basename(path))[1]
+
+    @property
+    def data(self) -> Optional[np.ndarray]:
+        """Image data in BGR HWC [0; 255] (float) format"""
+
+        if not self.has_data:
+            return None
+
+        data = self.__data()
+
+        if self._size is None and data is not None:
+            if not 2 <= data.ndim <= 3:
+                raise MediaShapeError("An image should have 2 (gray) or 3 (bgr) dims.")
+            self._size = tuple(map(int, data.shape[:2]))
+        return data
+
+    def save(
+        self,
+        fp: Union[str, io.IOBase],
+        ext: Optional[str] = None,
+        crypter: Crypter = NULL_CRYPTER,
+    ):
+        cur_path = osp.abspath(self.path) if self.path else None
+        cur_ext = self.ext
+        new_ext = self._get_ext_to_save(fp, ext)
+        if isinstance(fp, str):
+            os.makedirs(osp.dirname(fp), exist_ok=True)
+
+        if cur_path is not None and osp.isfile(cur_path):
+            if cur_ext == new_ext:
+                copyto_image(src=cur_path, dst=fp, src_crypter=self._crypter, dst_crypter=crypter)
+            else:
+                save_image(fp, self.data, ext=new_ext, crypter=crypter)
         else:
-            save_image(path, self.data, crypter=crypter)
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), cur_path)
 
     def set_crypter(self, crypter: Crypter):
         super().set_crypter(crypter)
-        if isinstance(self._data, lazy_image):
-            self._data._crypter = crypter
+        if isinstance(self.__data, lazy_image):
+            self.__data._crypter = crypter
+
+
+class ImageFromData(FromDataMixin, Image):
+    def save(
+        self,
+        fp: Union[str, io.IOBase],
+        ext: Optional[str] = None,
+        crypter: Crypter = NULL_CRYPTER,
+    ):
+        data = self.data
+        if data is None:
+            raise ValueError(f"{self.__class__.__name__} is empty.")
+        new_ext = self._get_ext_to_save(fp, ext)
+        if isinstance(fp, str):
+            os.makedirs(osp.dirname(fp), exist_ok=True)
+        save_image(fp, data, ext=new_ext, crypter=crypter)
+
+
+class ImageFromNumpy(ImageFromData):
+    def __init__(
+        self,
+        data: Union[Callable[[], bytes], bytes],
+        *args,
+        **kwargs,
+    ):
+        super().__init__(data=data, *args, **kwargs)
 
     @property
-    def path(self) -> str:
-        """Path to the media file"""
-        return self._path
+    def data(self) -> Optional[np.ndarray]:
+        """Image data in BGR HWC [0; 255] (float) format"""
+
+        data = super().data
+
+        if isinstance(data, np.ndarray):
+            data = data.astype(np.float32)
+        if self._size is None and data is not None:
+            if not 2 <= data.ndim <= 3:
+                raise MediaShapeError("An image should have 2 (gray) or 3 (rgb) dims.")
```

Comment:
could you check to support RGB(A)? That means, we may be possible to support `n.dim == 4`

Reply:
It was fixed in https://github.com/openvinotoolkit/datumaro/pull/921/commits/e19ad25b3abdf7c8103198a671e502e1f20d71d3